mod-source-record-manager-server/src/main/java/org/folio/verticle/DataImportJournalBatchConsumerVerticle.java

```diff
@@ -0,0 +1,425 @@
+          // Return empty Optional to skip this record and continue processing
+          return Optional.<Bundle>empty();
+        }
+      })
+      .filter(Optional::isPresent)
+      .map(Optional::get)
+      .flatMapSingle(bundle -> createJournalRecords(bundle)
+        .map(records -> Pair.of(Optional.of(bundle), records))
+        .onErrorReturnItem(Pair.of(Optional.empty(), Collections.emptyList()))
+      );
+  }
+
+  private Flowable<Pair<Optional<Bundle>, Collection<BatchableJournalRecord>>> listenEventBusMessages() {
+    if (eventBusConsumer == null) {
+      throw new IllegalStateException("journal eventbus consumer not initialized");
+    }
+
+    return eventBusConsumer
+      .bodyStream()
+      .toFlowable()
+      // Flatten the iterable list of messages
+      .flatMapIterable(list -> list)
+      // Window the messages in 2-second intervals, with a maximum of MAX_NUM_EVENTS per window
+      .window(2, TimeUnit.SECONDS, scheduler, MAX_NUM_EVENTS, true)
+      .flatMap(window -> window
+        // Group messages by tenant ID
+        .groupBy(msg -> msg.getJournalRecord().getTenantId())
+        .flatMapSingle(Flowable::toList)
+        .map(journalRecords -> Pair.of(Optional.empty(), journalRecords))
+      );
+  }
+
+  private Completable saveJournalRecords(ConnectableFlowable<Pair<Optional<Bundle>, Collection<BatchableJournalRecord>>> flowable) {
+    Completable completable = flowable
+      // Filter out pairs with empty records
+      .filter(pair -> !pair.getRight().isEmpty())
+      // Group records by tenant ID
+      .groupBy(pair -> {
+        Optional<BatchableJournalRecord> first = pair.getRight().stream().findFirst();
+        return first.map(journalRecord -> journalRecord.getJournalRecord().getTenantId());
+      })
+      // Process each group of records
+      .flatMapCompletable(groupedRecords -> groupedRecords.toList()
+        .flatMapCompletable(pairs -> {
+            try {
+              // Map and collect journal records, setting deterministic identifiers
+              Collection<JournalRecord> journalRecords = pairs
+                .stream()
+                .flatMap(pair -> pair.getRight().stream().map(BatchableJournalRecord::getJournalRecord))
+                .map(this::setDeterministicIdentifer)
+                .toList();
+              // If no records or tenant ID is missing, complete without action
+              if (journalRecords.isEmpty() || groupedRecords.getKey().isEmpty()) return Completable.complete();
+
+              LOGGER.info("saveJournalRecords:: Saving {} journal record(s) for tenantId={}", journalRecords.size(), groupedRecords.getKey().get());
+              // Save the batch of journal records and handle the response
+              return AsyncResultCompletable.toCompletable(completionHandler -> batchJournalService.saveBatchWithResponse(
+                journalRecords,
+                groupedRecords.getKey().get(),
```

Comment:
found that looks like we must not pass tenantId from the journal record to the method for saving journal records. Because journalRecord.tenantId field is used to keep id of tenant where entity/record is located over which action has been applied. This field is mostly used to keep central tenant id of the instance/record action on which action has been applied during data import initiated on the member tenant to ensure appearance of shared instance marc record and proper work of link to the updated shared instance on the data import logs pages. If we pass the tenantId from the journal record instead from the headers during a shared instance update by import from a member tenant, it will result in some journal records being saved to the central tenant DB schema instead of the member tenant schema, causing to the loss of logs data for the import job on the member tenant.
In the case of shared instance update by import from a member tenant, the tenantId in the headers will contain the member tenant id, and the journal record for the shared instance update will contain the tenantId of the central tenant.
Maybe we can add tenantId field to the BatchableJournalRecord class to keep there always tenant id from the headers, and then use tenantId from the BatchableJournalRecord to group journal records by tenant id before saving.
In this way, when passing groupedRecords key as a tenantId for journal record saving, we will always pass the tenantId from the headers.

Reply:
I have updated the code, please check it out!

Reply:
thank you!